src/device-manager/java/src/nl/Weave/DataManagement/WDMClientImpl.java

```diff
@@ -0,0 +1,213 @@
+/*
+
+    Copyright (c) 2019 Google, LLC.
+    All rights reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+package nl.Weave.DataManagement;
+
+import android.os.Build;
+import android.util.Log;
+
+import java.util.EnumSet;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+
+public class WDMClientImpl implements WDMClient
+{
+    public WDMClientImpl()
+    {
+        init();
+        mCompHandler = null;
+    }
+
+    @Override
+    public void connect(long deviceMgrPtr)
+    {
+        mWDMClientPtr = newWDMClient(deviceMgrPtr);
+    }
+
+    @Override
+    public void close()
+    {
+        if (mTraitMap != null)
```

Comment:
This can be simplified to:

```
if (mTraitMap != null) {
  for (GenericTraitUpdatableDataSink dataSink : mTraitMap.values()) {
    dataink.close();
  }
}
```

Note that the null check on the dataSink can be avoided by not inserting null into the map (instead just removing the key -- will expand on that below).

Reply:
If multiple data sink are being deallocated, and WDMClient is being deallocated by java garbage collection in parallel, the mTraitMap may be updated dynamically, the close may execute on the one that has been deallocated, the crash may happen.

Reply:
I don't think that could be an issue here. If this WDMClientImpl is holding a reference to the GenericTraitUpdatableDataSink, then they will not be garbage collected. So by definition of holding a reference to them here you should not run into any concurrency issues with GC.

Reply:
ok, thanks, done